cpctools/lib/libdsk/man/dskform.1
.\" -*- nroff -*-
.\"
.\" dskform.1: dskform man page
.\" Copyright (c) 2002 Darren Salt
.\" Copyright (c) 2005 John Elliott
.\"
.\" This library is free software; you can redistribute it and/or modify it
.\" under the terms of the GNU Library General Public License as published by
.\" the Free Software Foundation; either version 2 of the License, or (at
.\" your option) any later version.
.\"
.\" This library is distributed in the hope that it will be useful, but
.\" WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library
.\" General Public License for more details.
.\"
.\" You should have received a copy of the GNU Library General Public License
.\" along with this library; if not, write to the Free Software Foundation,
.\" Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA
.\"
.\" Author contact information:
.\" John Elliott: email: jce@seasip.demon.co.uk
.\"
.TH dskform 1 "20 January, 2006" "Version 1.1.9" "Emulators"
.\"
.\"------------------------------------------------------------------
.\"
.SH NAME
dskform - floppy or image file formatter for emulated machines
.\"
.\"------------------------------------------------------------------
.\"
.SH SYNOPSIS
.PD 0
.B dskform
.RI [ "-format FORMAT" ]
.RI [ "-type TYPE" ]
.RI [ "-side SIDE" ]
.RI [ "-comp COMP" ]
.RI [ "-retry COUNT" ]
.I DISKIMAGE
.P
.PD 1
.\"
.\"------------------------------------------------------------------
.\"
.SH DESCRIPTION
Dskform is a formatter which formats floppy discs or disc images for use with
emulated machines such as the ZX Spectrum +3, Amstrad CPC, and various Acorn
machines. It does
.B not
create a filesystem on the disc or image.
.\"
.\"------------------------------------------------------------------
.\"
.SH OPTIONS
.TP
.B -format FORMAT
Specifies a disc format to use. See the section DISK FORMATS for more
information.

.TP
.B -type TYPE
Determines which driver is to be used to format the disc or image.
.RS
.TP 8
.B auto
Select according to the disc image file. This is the default.
.TP
.B dsk
Use the DSK (CPCEmu format) image driver.
.TP
.B edsk
Use the extended version of the DSK format.
.TP
.B floppy
Use the floppy driver.
.TP
.B myz80
Use the hard disk (MYZ80 format) image driver.
(This format cannot be autodetected.)
.TP
.B cfi
Use the CFI (DOS fdcopy format) image driver.
(This format cannot be autodetected.)
.TP
.B apridisk
Use the ApriDisk image driver (used for archiving Apricot floppies).
(This format cannot be autodetected.)
.TP
.B raw
Use the raw driver.
.RE
.TP
.B -comp COMP
Compress the disc image that is being generated (has no
effect when formatting a real floppy disc).
.RS
.TP 8
.B sq
Huffman compressed (SQ / USQ).
.TP
.B gz
Gzipped (gzip / gunzip).
.RE

.TP
.B -retry COUNT
Set the number of times to attempt a format operation in case of error.

.TP
.B -side SIDE
Determines which side (0 or 1) of the disc is to be formatted.

.\"
.\" -----------------------------------------------------------------
.\"
.SH DISK FORMATS
.TP 20
.I pcw180
PCW or IBM; 180K. This is the default format.
.TP
.I cpcsys
CPC or +3 System
.TP
.I cpcdata
CPC or +3 Data
.TP
.I pcw720
PCW or IBM 720K
.TP
.I pcw1440
PCW16 or IBM 1440K
.TP
.I ibm160 ibm320
IBM 160K or 320K (CP/M-86 or DOSPlus)
.TP
.I ibm360
IBM 360K (DOSPlus)
.TP
.I ibm720 ibm1200 ibm1440
IBM 720K, 1200K or 1440K (144FEAT)
.TP
.I acorn160 acorn320
Acorn ADFS 'S' or 'M'
.TP
.I acorn640
Acorn ADFS 'L'
.TP
.I acorn800
Acorn ADFS 'D' and 'E'/'E+'
.TP
.I acorn1600
Acorn ADFS 'F'/'F+'
.TP
.I pcw800
PCW 800K (XCF2DD)
.TP
.I pcw200
PCW 200K (XCF2)
.TP
.I bbc100
BBC Micro DFS (single sided, 40 tracks)
.TP
.I bbc200
BBC Micro DFS (single sided, 80 tracks)

.\"
.\"------------------------------------------------------------------
.\"
.\".SH BUGS
.\"
.\"------------------------------------------------------------------
.\"
.\".SH SEE ALSO
.\"
.\"------------------------------------------------------------------
.\"
.\" `AUTHOR' here is deliberate...
.\"
.SH AUTHOR
John Elliott <jce@seasip.demon.co.uk>.
.PP
Darren Salt wrote the man pages.


cpctools/NEWS
2008/09/17:
	Now works with gcc-4.3


cpctools/lib/libdsk_Release/tools/dskform.c
/***************************************************************************
 *                                                                         *
 *    LIBDSK: General floppy and diskimage access library                  *
 *    Copyright (C) 2001-2  John Elliott <jce@seasip.demon.co.uk>          *
 *                                                                         *
 *    This library is free software; you can redistribute it and/or        *
 *    modify it under the terms of the GNU Library General Public          *
 *    License as published by the Free Software Foundation; either         *
 *    version 2 of the License, or (at your option) any later version.     *
 *                                                                         *
 *    This library is distributed in the hope that it will be useful,      *
 *    but WITHOUT ANY WARRANTY; without even the implied warranty of       *
 *    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU    *
 *    Library General Public License for more details.                     *
 *                                                                         *
 *    You should have received a copy of the GNU Library General Public    *
 *    License along with this library; if not, write to the Free           *
 *    Software Foundation, Inc., 59 Temple Place - Suite 330, Boston,      *
 *    MA 02111-1307, USA                                                   *
 *                                                                         *
 ***************************************************************************/

/* Example formatter */

#include <stdio.h>
#include <string.h>
#include "config.h"
#include "libdsk.h"
#include "utilopts.h"
#include "formname.h"

#ifdef CPM
#define AV0 "DSKFORM"
#else
#define AV0 argv[0]
#endif

static int retries = 1;

int do_format(char *outfile, char *outtyp, char *outcomp, int forcehead, dsk_format_t format);

int help(int argc, char **argv)
{
	fprintf(stderr, "Syntax: \n"
                "      %s { -format <format> } { -retry <count> }"
                " { -type <type> } { -side <side> } dskimage \n",
		AV0);
	fprintf(stderr,"\nDefault type is DSK.\nDefault format is PCW 180k.\n\n");
		
	fprintf(stderr, "eg: %s myfile.DSK\n"
                        "    %s -type floppy -format cpcsys -side 1 /dev/fd0\n", AV0, AV0);

	valid_formats();
	return 1;
}


static void report(const char *s)
{
        printf("%s\r", s);
        fflush(stdout);
}

static void report_end(void)
{
        printf("\r%-79.79s\r", "");
        fflush(stdout);
}



int main(int argc, char **argv)
{
	char *outtyp;
	char *outcomp;
	int forcehead;
	dsk_format_t format;

	if (argc < 2) return help(argc, argv);
        if (find_arg("--help",    argc, argv) > 0) return help(argc, argv);
        if (find_arg("--version", argc, argv) > 0) return version();
        ignore_arg("-itype", 2, &argc, argv);
        ignore_arg("-iside", 2, &argc, argv);
        ignore_arg("-icomp", 2, &argc, argv);
        ignore_arg("-otype", 2, &argc, argv);
        ignore_arg("-oside", 2, &argc, argv);
        ignore_arg("-ocomp", 2, &argc, argv);
        outtyp    = check_type("-type", &argc, argv); if (!outtyp) outtyp = "dsk";
        outcomp   = check_type("-comp", &argc, argv); 
        forcehead = check_forcehead("-side", &argc, argv);
	format    = check_format("-format", &argc, argv);
        retries   = check_retry("-retry", &argc, argv);
	if (format == -1) format = FMT_180K;
	args_complete(&argc, argv);

	return do_format(argv[1], outtyp, outcomp, forcehead, format);
}

static unsigned char spec169 [10] = { 0,    0, 40, 9, 2, 2, 3, 2, 0x2A, 0x52 };
static unsigned char spec180 [10] = { 0,    0, 40, 9, 2, 1, 3, 2, 0x2A, 0x52 };
static unsigned char spec200 [10] = { 0,    0, 40,10, 2, 1, 3, 3, 0x0C, 0x17 };
static unsigned char spec720 [10] = { 3, 0x81, 80, 9, 2, 1, 4, 4, 0x2A, 0x52 };
static unsigned char spec800 [10] = { 3, 0x81, 80,10, 2, 1, 4, 4, 0x0C, 0x17 };

int do_format(char *outfile, char *outtyp, char *outcomp, int forcehead, dsk_format_t format)
{
	DSK_PDRIVER outdr = NULL;
	dsk_err_t e;
	dsk_pcyl_t cyl;
	dsk_phead_t head;
	DSK_GEOMETRY dg;
	dsk_cchar_t fdesc;

	dsk_reportfunc_set(report, report_end);
	e = dsk_creat(&outdr, outfile, outtyp, outcomp);
	if (!e) e = dsk_set_retry(outdr, retries);
	if (!e && forcehead >= 0) e = dsk_set_forcehead(outdr, forcehead);
	if (!e) e = dg_stdformat(&dg, format, NULL, &fdesc);
	if (!e)
	{
		printf("Formatting %s as %s\n", outfile, fdesc);
		for (cyl = 0; cyl < dg.dg_cylinders; cyl++)
		{
		    for (head = 0; head < dg.dg_heads; head++)
		    {
			printf("Cyl %02d/%02d Head %d/%d\r", 
				cyl +1, dg.dg_cylinders,
			 	head+1, dg.dg_heads);
			fflush(stdout);

			if (!e) e = dsk_apform(outdr, &dg, cyl, head, 0xE5);
			if (e) break;	
		    }
		}
	
	}
	/* Create a disc spec on the first sector, if the format's recognised */
	if (!e)
	{
		unsigned char bootsec[512];
		unsigned int do_bootsec = 0;

		memset(bootsec, 0xE5, sizeof(bootsec));
		switch(format)
		{
			case FMT_180K:
				memcpy(bootsec, spec180, 10);
				do_bootsec = 1;
				break;
			case FMT_CPCSYS:
				memcpy(bootsec, spec169, 10);
				do_bootsec = 1;
				break;
			case FMT_CPCDATA:
				break;
			case FMT_200K:
				memcpy(bootsec, spec200, 10);
				do_bootsec = 1;
				break;
			case FMT_720K:
				memcpy(bootsec, spec720, 10);
				do_bootsec = 1;
				break;
			case FMT_800K:
				memcpy(bootsec, spec800, 10);
				do_bootsec = 1;
				break;
			default:
				break;
		}
		if (do_bootsec) e = dsk_lwrite(outdr, &dg, bootsec, 0);
	}
	printf("\r                                     \r");
	if (outdr) dsk_close(&outdr);
	if (e)
	{
		fprintf(stderr, "%s\n", dsk_strerror(e));
		return 1;
	}
	return 0;
}


